c++/mpi/window.hpp

```diff
@@ -68,83 +94,156 @@ namespace mpi {
         MPI_Win_free(&win);
+      } else {
+        if (is_local) {
+          std::free(base_local);
+          base_local = nullptr;
+        }
       }
     }
 
     /// Synchronization routine in active target RMA. It opens and closes an access epoch.
     void fence(int assert = 0) const noexcept {
-      MPI_Win_fence(assert, win);
+      if (has_env) {
+        MPI_Win_fence(assert, win);
+      }
     }
 
     /// Complete all outstanding RMA operations at both the origin and the target
     void flush(int rank = -1) const noexcept {
-      if (rank < 0) {
-        MPI_Win_flush_all(win);
-      } else {
-        MPI_Win_flush(rank, win);
+      if (has_env) {
+        if (rank < 0) {
+          MPI_Win_flush_all(win);
+        } else {
+          MPI_Win_flush(rank, win);
+        } 
       }
     }
 
     /// Synchronize the private and public copies of the window
     void sync() const noexcept {
-      MPI_Win_sync(win);
+      if (has_env) {      
+        MPI_Win_sync(win);
+      }
     }
 
     /// Starts an RMA access epoch locking access to a particular or all ranks in the window
     void lock(int rank = -1, int lock_type = MPI_LOCK_SHARED, int assert = 0) const noexcept {
-      if (rank < 0) {
-        MPI_Win_lock_all(assert, win);
-      } else {
-        MPI_Win_lock(lock_type, rank, assert, win);
+      if (has_env) {
+        if (rank < 0) {
+          MPI_Win_lock_all(assert, win);
+        } else {
+          MPI_Win_lock(lock_type, rank, assert, win);
+        }
       }
     }
 
     /// Completes an RMA access epoch started by a call to lock()
     void unlock(int rank = -1) const noexcept {
-      if (rank < 0) {
-        MPI_Win_unlock_all(win);
-      } else {
-        MPI_Win_unlock(rank, win);
+      if (has_env) {
+        if (rank < 0) {
+          MPI_Win_unlock_all(win);
+        } else {
+          MPI_Win_unlock(rank, win);
+        }
       }
     }
 
     /// Load data from a remote memory window.
     template <typename TargetType = BaseType, typename OriginType>
     std::enable_if_t<has_mpi_type<OriginType> && has_mpi_type<TargetType>, void>
     get(OriginType *origin_addr, int origin_count, int target_rank, MPI_Aint target_disp = 0, int target_count = -1) const noexcept {
-      MPI_Datatype origin_datatype = mpi_type<OriginType>::get();
-      MPI_Datatype target_datatype = mpi_type<TargetType>::get();
       int target_count_ = target_count < 0 ? origin_count : target_count;
-      MPI_Get(origin_addr, origin_count, origin_datatype, target_rank, target_disp, target_count_, target_datatype, win);
-    };
+      if (has_env) {
+        MPI_Datatype origin_datatype = mpi_type<OriginType>::get();
+        MPI_Datatype target_datatype = mpi_type<TargetType>::get();
+        MPI_Get(origin_addr, origin_count, origin_datatype, target_rank, target_disp, target_count_, target_datatype, win);
+      } else {
+        if ((sizeof(OriginType) * origin_count) >= (sizeof(TargetType) * target_count_)) {
```

Comment:
```suggestion
        if ((sizeof(OriginType) * origin_count) >= (sizeof(TargetType) * target_count_) + target_disp) {
```
Assumes that `target_disp` is in bytes. (Check with the standard maybe?)